text for slack notifs (#948)
returntocorp/semgrep-docs: docs/semgrep-code/notifications.md

**Policies **typically correspond to the different Rule board columns. For example, you can choose to receive notifications only for findings generated by rules from the Rule Board's Blocking column.

:::info PREREQUISITE
You must be a Slack **Workspace Owner** to set up the Semgrep Slack app. 
:::

#### Installing Semgrep Slack App

To install the Semgrep Slack app, follow these steps:

1. Sign in to your [Semgrep Cloud Platform](semgrep.dev/login) account, and then go to **Settings** > **[Integrations](https://semgrep.dev/orgs/-/settings/integrations)**.
2. On the **[Integrations](https://semgrep.dev/orgs/-/settings/integrations)** page, click **Add Integration** (or **Setup First Integration** if this is your first integration), and then select **Slack**.
3. Click **Allow**.

#### Setting up notifications for findings in Slack

To set up or subscribe to notifications for findings in your Slack workspace, perform the following steps:

1. In your Slack workspace, find or create a channel for Semgrep notifications.

> 2. In the selected Slack channel, enter the following slash command: `/semgrep_subscribe`.

Reviewer: Should we be including this just as a heads up or FYI?

> If customers select >1 filter, i.e. project, policy, product; it's more like an **AND** instead of **OR**. Which means, a user will only receive notifications when all of those conditions are met.

Author: Thanks for this heads up! I'll add a section in troubleshooting "Not receiving any findings".

> 3. **Optional**: Enter the name of a specific project after `/semgrep_subscribe` to receive findings for that specific project only, for example, `/semgrep_subscribe purposely-vulnerable-go-repo`.

Reviewer: Should we make a note here that if this filter is not added, they will receive notifications for all the projects by default?

Author: Updated with this suggestion!

> 4. Choose an organization in the list under **Select target organization**. The dialog box expands with additional options.

Author: I added a note about the correct formatting (ACCOUNT_NAME/REPO_NAME). In case of failure, they'll see the message in the Slack app itself, so we don't have to mention it :)

    ![Semgrep Slack app dialog box for subscribing to notifications](/img/semgrep-slack-subscribe.png "Semgrep Slack app dialog box for subscribing to notifications")
4. **Optional**: For Semgrep users that receive both Semgrep Code findings and Semgrep Supply Chain vulnerabilities, you can select **target scan types** to subscribe to either Semgrep Code, Semgrep Supply Chain, or both.

> 5. **Optional:** Select any number of policies to receive findings for under the **Selected Policies** field. Note that selecting all policies, including the Monitor policy, can potentially result in a noisy message.

Reviewer: Should we make a note here that if this filter is not selected, they will receive notifications for all the policies by default?

Author: Added this bit in!

2. Click **Semgrep**. The Semgrep app appears as a button on the sidebar.

#### Uninstalling the Semgrep Slack App from your Slack workspace

:::caution
This removes **all** Semgrep notifications in **all** channels in your Slack workspace. 
:::

To uninstall the Semgrep Slack App entirely from your Slack workspace, perform the following steps:

1. Sign in to your [Semgrep Cloud Platform](semgrep.dev/login) account, and then go to **Settings** > **[Integrations](https://semgrep.dev/orgs/-/settings/integrations)**.
2. On the **[Integrations](https://semgrep.dev/orgs/-/settings/integrations)** page, find the Slack integration you want to remove.
3. Click **Remove integration** > **Remove**.

#### Troubleshooting

#### Receiving `dispatch_failed` error

There are many possible causes for this error. Possible fixes include:

* Re-entering your last command or operation after a few minutes.
* Uninstalling then reinstalling your Semgrep Slack integration.

#### Receiving `operation_timeout` error

This error occasionally appears due to connection or service issues. To fix this issue, retry your last command or operation after a few minutes.

> 

Author: Added and done! Thanks for pointing this out.

<!-- When referring to the Slack integration, let's consistently call it the "Semgrep Slack app" (app in lower case). This is consistent with the terminology used by Slack (they refer to these integrations as apps in their product pages. It also makes sure that the user knows what "app" we're referring to.

Prefer "Semgrep Slack app" over "Semgrep Slack integration" or "Slack integration" or any other variant. -->

<!-- <ProcedureIntegrateSlack /> -->

The Semgrep Slack app enables Semgrep Cloud Platform to notify you of new findings after every scan. By receiving notifications within your Slack workspace, developers and security engineers are able to see findings without leaving their current environment. This can lessen the friction between detecting a finding, triaging it, and resolving it.

You can select which channels in your Slack workspace will receive findings. You can also choose to receive findings only for certain repositories (projects) or certain **policies**. 

**Policies **typically correspond to the different Rule board columns. For example, you can choose to receive notifications only for findings generated by rules from the Rule Board's Blocking column.

> :::info PREREQUISITE

Reviewer 2: Optional suggestion: These headers are all rendered as capitalized, this is just source code consistency edit (feel free to dismiss). 
```suggestion
:::info Prerequisite
```

You must be a Slack **Workspace Owner** to set up the Semgrep Slack app. 
:::

#### Installing Semgrep Slack App

To install the Semgrep Slack app, follow these steps:

> 1. Sign in to your [Semgrep Cloud Platform](https://semgrep.dev/login) account, and then go to **Settings** > **[Integrations](https://semgrep.dev/orgs/-/settings/integrations)**.

Reviewer 2: Optional: Maybe we can shorten the first steps in procedures.
```suggestion
1. In [Semgrep Cloud Platform](https://semgrep.dev/login), go to **Settings** > **[Integrations](https://semgrep.dev/orgs/-/settings/integrations)**.
```

2. On the **[Integrations](https://semgrep.dev/orgs/-/settings/integrations)** page, click **Add Integration** (or **Setup First Integration** if this is your first integration), and then select **Slack**.
3. Click **Allow**.

#### Setting up notifications for findings in Slack

To set up or subscribe to notifications for findings in your Slack workspace, perform the following steps:

1. In your Slack workspace, find or create a channel for Semgrep notifications.
2. In the selected Slack channel, enter the following slash command: `/semgrep_subscribe`.

> 3. **Optional**: Enter the name of a specific project after `/semgrep_subscribe` to receive findings for that specific project only, for example, `/semgrep_subscribe acme-corp/vulnerable-repo`. The project must be entered in the following format:

Reviewer 2: Do not make the "Optional" bold. It gives it more visual importance than other required steps.
```suggestion
3. Optional: Enter the name of a specific project after `/semgrep_subscribe` to receive findings for that specific project only, for example, `/semgrep_subscribe acme-corp/vulnerable-repo`. The project must be entered in the following format:
```

    <br />
    <code>/semgrep_subscribe <span className="placeholder">ACCOUNT_NAME/REPOSITORY_NAME</span></code>
4. Choose an organization in the list under **Select target organization**. The dialog box expands with additional options.
    ![Semgrep Slack app dialog box for subscribing to notifications](/img/semgrep-slack-subscribe.png "Semgrep Slack app dialog box for subscribing to notifications")

> 5. **Optional**: Set up additional filters.
>     1. For Semgrep users that receive both Semgrep Code findings and Semgrep Supply Chain vulnerabilities, you can select **target scan types** to subscribe to either Semgrep Code, Semgrep Supply Chain, or both.
>     2. Select any number of policies to receive findings for under the **Selected Policies** field. By default, you are subscribed to all policies, including the Monitor policy. This can potentially result in a noisy message.

Reviewer 2: Optional (matter of style we did not define previously, feel free to dismiss). See [style guide](https://developers.google.com/style/procedures). To make the list more concise, you don't have to use indentation:
```suggestion
5. Optional: For Semgrep users that receive both Semgrep Code findings and Semgrep Supply Chain vulnerabilities, you can select **target scan types** to subscribe to either Semgrep Code, Semgrep Supply Chain, or both.
6. Optional: Select any number of policies to receive findings for under the **Selected Policies** field. By default, you are subscribed to all policies, including the Monitor policy. This can potentially result in a noisy message.
```

Author: I tried both in an earlier commit and chose this because I want to make it clear we're applying filters. Also having so many "optional" texts clutters up the procedure. I'll take out the bold formatting though.

6. Click **Subscribe**. If you did not specify a project after `/semgrep_subscribe`, the channel is subscribed to findings from all your repositories in Semgrep Cloud Platform.
7. **Optional**: To set up Slack notifications for additional workspaces, repeat steps 1 to 6. The Semgrep Slack integration is set up on a per-workspace basis.

> You have successfully set up notifications for Semgrep findings. The Semgrep Slack app reports new findings after every scan. It does not report findings that were previously discovered.

Reviewer 2: Optional:
```suggestion
You have successfully set up notifications for Semgrep findings. The Semgrep Slack app reports new findings after every scan but does not report findings that were previously discovered.
```

:::tip Suggested workflow
In your Slack workspace, create separate channels for either policies, repositories (projects), or types of findings depending on your business or development need. This ensures that developers receive only findings that are relevant to them. 
:::

![Message from Semgrep Slack app consolidating findings after a scan](/img/slack-findings-message.png#bordered "Message from Semgrep Slack app consolidating findings after a scan")
_Figure 1_. A sample Slack message with Semgrep findings.

#### Removing notifications for findings in Slack

:::note
This operation removes or unsubscribes a channel from notifications. To uninstall the Semgrep Slack App, refer to [Uninstalling the Semgrep Slack App from your Slack workspace
](#uninstalling-the-semgrep-slack-app-from-your-slack-workspace).
:::

To remove or unsubscribe to notifications:

> 1. Enter the channel that will unsubscribe from Semgrep findings.

Reviewer 2: Specify where is the action happening (Slack). It is partially done by saying "channel", so only an optional suggestion.

Optional:
```suggestion
1. In Slack, enter the channel that you want to unsubscribe from Semgrep findings.
```

2. Type `/semgrep_unsubscribe`.
3. Select the target organization to unsubscribe from.
4. Click **Unsubscribe**.

You have unsubscribed from Semgrep finding notifications for that particular channel.

> #### Making changes to your Slack notifications

Reviewer 2: Optional: This header may sound too unspecific. Changes of what kind? Making a suggestion below, feel free to dismiss it.
```suggestion
#### Changing Slack notification settings
```

Or:
```suggestion
#### Changing your Slack notifications
```

> #### Making changes to your Slack notifications

Reviewer 2: Optional: This header may sound too unspecific. Changes of what kind? Making a suggestion below, feel free to dismiss it.
```suggestion
#### Changing Slack notifications settings
```

Or (this might be even less specific though :D ):
```suggestion
#### Changing your Slack notifications
```

You can customize your notification settings at any time through the **Semgrep App Home** in your Slack workplace.

To view the Semgrep App Home:

1. In your Slack workspace,  click **+ Add apps**, which can be found in the sidebar under the **Apps** header.
2. Click **Semgrep**. The Semgrep app appears as a button on the sidebar.

To change the settings:

1. In your Slack workspace, click **Semgrep** under **Apps** in the Slack sidebar. This displays the **Semgrep App Home**.
2. Click the **three-dot menu** of the channel to update.
3. Click **Manage filters**.

#### Uninstalling the Semgrep Slack App from your Slack workspace

:::caution
This removes **all** Semgrep notifications in **all** channels in your Slack workspace. 
:::

To uninstall the Semgrep Slack App entirely from your Slack workspace, perform the following steps:

> 1. Sign in to your [Semgrep Cloud Platform](semgrep.dev/login) account, and then go to **Settings** > **[Integrations](https://semgrep.dev/orgs/-/settings/integrations)**.

Reviewer 2: Optional: If you chose the shortened procedure above.
```suggestion
1. In [Semgrep Cloud Platform](https://semgrep.dev/login), go to **Settings** > **[Integrations](https://semgrep.dev/orgs/-/settings/integrations)**.
```

2. On the **[Integrations](https://semgrep.dev/orgs/-/settings/integrations)** page, find the Slack integration you want to remove.
3. Click **Remove integration** > **Remove**.

#### Troubleshooting

#### Not receiving any findings

> The following list describes possible ways to troubleshoot findings not appearing in your Slack workspace:

Reviewer 2: Optional:
```suggestion
The following list describes possible ways to troubleshoot findings that do **not** appear in your Slack workspace:
```

> * Check that you have successfully set up your notifications.
> * Check that your most recent scan has findings to send.

Reviewer 2: ```suggestion
* Check if you have successfully set up your notifications.
* Check if your most recent scan has findings to send.
```

* Check your filters.

##### Check notifications

To check that your notifications are set up, receive a test message from Semgrep:

1. In your Slack workspace, click **Semgrep** under **Apps** in the Slack sidebar.
2. Click the **three-dot menu** > **Send Test Notification**.

##### Check your filters

If you have set up any filter, such as filtering for a specific policy or project, all conditions of that filter must be present for the notificatin to be sent. Review your filters by following the steps in [Making changes to your Slack notifications](#making-changes-to-your-slack-notifications).

#### Permissions not up-to-date

You may receive a message from Semgrep Slack app stating that your token does not have up-to-date permissions. Clicking the link provided in the message to update the permissions typically resolves this issue.

> However, if after updating the token, you still receive the same message, perform the following steps to revoke then refresh your access token:

Author: I'll go with "Revoke and refresh" so that the process is treated as a group. ("Then" makes it sound like "refresh" is a separate procedure.)

> However, if after updating the token, you still receive the same message, perform the following steps to revoke then refresh your access token:

Reviewer 2: @s-santillan good idea!
